### PR TITLE
Rust parser for SIP protocol v2

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3084

Describe changes:
This patchset consists of a parser, logging and detection keywords for SIP protocol.
All the SIP fields of a request and response are parsed, and they all are logged except for header fields.
The logging module provide basic logging, which means that extended or custom logging is not provided yet.
The following new keywords are implemented as sticky buffer:

- sip.method
- sip.uri
- sip.protocol
- sip.stat_code
- sip.stat_msg
- sip.request_line
- sip.response_line

Updates:
- `#ifdef HAVE_RUST` removed

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- PR glongo: https://buildbot.openinfosecfoundation.org/builders/glongo/builds/218
- PR glongo-pcap: https://buildbot.openinfosecfoundation.org/builders/glongo-pcap/builds/82

### CODE REVIEW

File: doc/userguide/rules/sip-keywords.rst [hunk 1]
```diff
@@ -0,0 +1,179 @@
+SIP Keywords
+============
+
+The SIP keywords are impletemented as sticky buffers can be used to match on fields in SIP messages.
```
Comment (norg) [inline doc/userguide/rules/sip-keywords.rst]: The SIP keywords are implemented as sticky buffers and can be used to match on fields in SIP messages.

File: src/output-json-sip.c [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Pierre Chifflier <chifflier@wzdftpd.net>
```
Comment (jasonish) [inline src/output-json-sip.c]: Did Pierre write this file or did you?

File: rust/src/sip/log.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+use sip::sip::{SIPState, SIPTransaction};
+
+#[no_mangle]
+pub extern "C" fn rs_sip_log_json(_state: &mut SIPState, tx: &mut SIPTransaction) -> *mut JsonT
+{
+    let js = Json::object();
+
+    match tx.request {
+        Some(ref req) => {
+            js.set_string("method", &format!("{}", req.method));
```
Comment (jasonish) [inline rust/src/sip/log.rs]: (minor) Format is required for any of these.  `js.set_string("method", &req.method)` is a little cleaner.

File: rust/src/sip/parser.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+                if !is_header_value(b) {
+                    return Err(Err::Incomplete(Needed::Size(1)));
+                }
+                end_pos = idx + 1;
+            }
+        }
+        idx += 1;
+    }
+    Ok((&b""[..], buf))
+}
```
Comment (jasonish) [inline rust/src/sip/parser.rs]: There should be unit tests here to test all the parsers.

File: rust/src/sip/detect.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+                *buffer_len = m.len() as u32;
+                return 1;
+            }
+        }
+        None => {}
+    }
+
+    *buffer = ptr::null();
+    *buffer_len = 0;
+
```
Comment (jasonish) [inline rust/src/sip/detect.rs]: Minor style comment. The following is more idiomatic when the `None` case is empty:
```
if let Some(ref r) = tx.request {
}
```

File: src/output-json-sip.c (hunk 2 above)
Comment (glongo) [reply src/output-json-sip.c]: I wrote it, just bad copy/paste.

**Comment by glongo** [comment]
Thanks Jason for the review, I will submit a new PR asap.

**Pull request closed without merging**